### PR TITLE
Add/Remove filters via the column header context menu

### LINKED ISSUE #2232: Allow users to add/remove filter conditions via the column header menu
## Problem
Currently, users can only apply filters to specific columns from the filter options in the table toolbar. It would be convenient to have filter options available at the column header level. 

## Proposed Solution
We propose adding a filter option to the column header menu to make it easier for users to filter specific columns. The option would open the current filter popup and have the selected column pre-selected. 

[Figma design](https://www.figma.com/file/xHb5oIqye3fnXtb2heRH34/Styling?node-id=1304%3A7357&t=WotuvB0jSNIg0eab-0)

image

## Additional Questions
From [Matrix Discussion](https://matrix.to/#/!xInuTkBwjXZXYatIlm:matrix.mathesar.org/$X5V9YArIFwaVVB7q341a_u1fTx_IhlNNWrfejm2g2Zg?via=matrix.mathesar.org&via=matrix.org)
- When a filter condition is already applied, would it be worth keeping the "Filter Column" option in the context menu, allowing the user to add another filter condition too?
- And if more the one filter condition is applied, we'll need to consider what behavior the "Remove Filter" action should take.

### PR DESCRIPTION
**Pull request by seancolsen**
Fixes #2232

## Demo

https://user-images.githubusercontent.com/42411/230149696-ed713d92-7b4e-4f52-8ac6-87f542d9e8e3.mp4

## Details

- The context menu text varies according to the number of existing filter entries for the column. You can see this in the demo video. "Filter Column" vs "Add Filter" and also "Remove Filter" vs "Remove 2 Filters"

- When adding a new filter, the filter dropdown is opened and the value input is focused. If the value input is a LinkedRecordInput, then the record selector is opened.

## Technical notes

This PR requires quite a bit of imperative logic, which we generally try to avoid and thus don't have a whole lot of in our codebase. To handle this logic, I've introduced two new patterns that are worth mentioning.

- I created an `ImperativeFilterController` which is set in context on the Table Page. It uses a simple `MessageBus` system to pass messages between components. The context menu which triggers the filter mutations _sends_ messages through the bus, and other components within the tree _listen_ to those messages to do things like open the filtering dropdown, add the filter entry, and activate the input. I'm doubtful that there would be a way to handle this feature declaratively, but I'm open to suggestions for other approaches which could be cleaner or easier to understand.

- In `LinkedRecordInput.svelte`, I'm assigning a function within the component to a method on a DOM node bound within the component. I don't think we're doing this anywhere else in our codebase. This is the best way I could come up with to handle the need to open the record selector.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by pavish** [comment]
@seancolsen Can you resolve the merge conflicts in this PR?

**Comment by seancolsen** [comment]
@pavish done

**Review by pavish — CHANGES_REQUESTED** [review]
@seancolsen 

I'm okay with the imperative logic here, I don't see a simpler way to make it declarative.

**Suggestion**:

A more readable & maintainable approach would be to move all logic within `Filter.svelte` into a headless `FilterDropdownController.svelte`, which supplies apis and stores for children to make use of. A similar approach is followed in `ListBox`.

Something like (This is only to represent the approach):
```
<FilterDropdownController let:api let:stores>
    <ButtonMenuItem on:click={() api.addFilter(columnId)}>
      ...
    </ButtonMenuItem>
    ...
   <FilterDropdown {api} {stores} /> <!-- pass api or access via context within the dropdown -->
</FilterDropdownController>
```

Within `FilterDropdownController`, we'd have all the methods that are currently in `Filter.svelte`. `Filter.svelte` will not have any local states or functions and will purely be a UI only component and call the functions from `FilterDropdownController` via context.

File: mathesar_ui/src/utils/MessageBus.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    this.listeners.add(listener);
+    return () => {
+      this.listeners.delete(listener);
+    };
+  }
+
+  send(message: Message) {
+    this.listeners.forEach((l) => l(message));
+  }
+}
```
[the remark this answers is not in the document]
Comment (seancolsen) [reply mathesar_ui/src/utils/MessageBus.ts]: Thanks for pointing that out. I hadn't considered `EventHandler`. I pushed 3508a0802e70b91e255cc921684e9bd3bede1731 to utilize it.

File: mathesar_ui/src/pages/table/ImperativeFilterController.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    this.activateLastFilterInput = new MessageBus();
+  }
+
+  async beginAddingNewFilteringEntry(columnId: number) {
+    this.openDropdown.send();
+    await tick();
+    this.addFilter.send(columnId);
+    await tick();
+    this.activateLastFilterInput.send();
+  }
```
[the remark this answers is not in the document]
Comment (seancolsen) [reply mathesar_ui/src/pages/table/ImperativeFilterController.ts]: I'm not concerned about this. While I don't have a deep enough understanding of the svelte internals to offer _proof_ that we're okay, I'll offer these to arguments to suggest that we're okay _beyond a reasonable doubt_:

- **Empirical evidence** In this PR, if you remove `tick()`, the feature does not work. With `tick()` the feature _does_ work, and does not throw any errors.

    Again this is not _proof_ that it never would through an error, if subjected to an edge case &mdash; but I think it lends some reasonable confidence to this usage.

- **Lack of contradictory evidence:** The Svelte documentation presents `tick` within its [Runtime API](https://svelte.dev/docs#run-time) section right alongside other utility function like `setContext` that we use heavily outside of components.

    Hypothetically, if `tick` were to behave problematically when called from a function defined outside a component, then I would expect:

    - (ideally) the svelte API _design_ to make that sort of footgun self evident or impossible
    - (or at the very least) the svelte docs to mention that sort of footgun.
    - (or maybe even) a web search for "svelte tick outside component" to return some meaningful results

    ...but none of the above seem to be true, which leads me to believe that it's _not_ problematic.

- **Precedent**: We've had `tick()` inside `portalToWindowFooter.ts` for ~5 months without issue.

**Comment by seancolsen** [comment]
@pavish this is ready for another look.

I agree that `Filter.svelte` is ripe for some refactoring. I'm not 100% clear on the changes you're suggesting, but from what I _do_ understand of your suggestions, I like. I'll also say that I very much like the design of `ListBox`. Let's just take up this refactoring work later. Even in the vein of refactors, I think we have higher priorities at the moment.

File: mathesar_ui/src/pages/table/ImperativeFilterController.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pavish) [reply mathesar_ui/src/pages/table/ImperativeFilterController.ts]: This sounds good to me.

**Review by pavish — APPROVED** [review]
Looks good!

**Pull request merged**